css-shared-element-transitions/index.bs

```diff
@@ -692,37 +698,43 @@ The {{SameDocumentTransition/finished}} [=getter steps=] are to return [=this's=
     </pre>
 </div>
 
-## Abandon the page transition ## {#abandon-the-page-transition-algorithm}
+## Skip the page transition ## {#skip-the-page-transition-algorithm}
 
 <div algorithm>
-    To <dfn>abandon the page transition</dfn> for {{SameDocumentTransition}} |transition|
-    with |error|:
+    To <dfn>skip the page transition</dfn> for {{DOMTransition}} |transition| with reason |reason|:
+
+    1. [=Assert=]: |transition| has not [=DOMTransition/concluded=].
+
+    1. If |transition|'s [=DOMTransition/DOM change callback called=] is false, then [=call the DOM change callback=] of |transition|.
```

Comment:
I think we need to clarify timings of things. These callbacks are asynchronous, and I'm not sure we should just call the callback and forget about it anytime a transition is aborted, since that has a potential of calling several callbacks before any one of the is finished, right? 

I was under the impression that we should queue the callback into some queue that is always calling the next callback, _waits for the returned promise to settle_, and then call the next callback. That may have a potential of some sort of a deadlock i'm sure, but that would be a developer error.

WDYT?

Reply:
I brought this up in https://github.com/WICG/shared-element-transitions/issues/175#issuecomment-1220817698

> While writing (the PR), I think we got something wrong in our plan:
> 
> > We need the pending callbacks to run before capturing outgoing state for the new transition.
> 
> I think this violates our "keep the transition and navigation separate" rule. Something other than our API should be in charge of queuing callbacks that should be queued.
> 
> The problem comes if a callback takes a long time to resolve. We would either:
> 
> * Queue indefinitely, meaning one bad callback can prevent all future queued DOM changes on the page.
> * Timeout. But, in that case, the callback could still make DOM changes, so we're still allowing these callbacks to run in parallel.
> 
> Neither of these are good.
> 
> The navigation API doesn't perform any queuing, it relies on developers to manage that themselves. It feels like we should follow the same model.

Reply:
I guess the key question is: is it our responsibility to queue these things, or is that something an asynchronous DOM system will already be doing, and we'd just get in the way by doing it again.

Reply:
Yeah, I agree that this is the question. My impression from our discussion was that the API would be responsible for serializing the callbacks. It seems like that may be problematic and does not align with navigation API. We likely should discuss this briefly to understand that timing of these callbacks, specifically their completion, would then depend on the callback duration itself, meaning it's fairly arbitrary

Reply:
Yes, in our discussion we were definitely set on queuing. I'm just not sure we thought the deadlock thing through. Or, at least, I hadn't.

Reply:
Could you give an example of a case where this could result in deadlocks? If the callbacks are being executed in the order createTransition is called, why would the first callback need to depend on any operation in the second callback.

@vmpstr and I discussed 3 options for this:

1. Execute the callbacks concurrently and while all steps for the transition are being executed. Since these callbacks are updating the DOM we could end up snapshotting a random frame and/or have the callbacks update the DOM midway through a transition. That seems pretty bad.
2. Execute the callbacks concurrently but wait for promises to settle before snapshotting. This avoids the latency of everything executing serially but the execution within each of them will interleave. But it ensures we're in a stable state before initiating capture and rest of the algorithm steps.
3. Execute the callbacks serially, waiting for each's promise to settle. And then initiate the transition. Sounds like this has the possibility for deadlocks which I'd like to understand better.

The first one seems pretty bad since we'll be running callbacks that are intended to update the DOM during the transition. Both 2 or 3 would be fine. If the developer doesn't like the default they can always build a queue in user land to switch to the other behaviour.

Reply:
> Could you give an example of a case where this could result in deadlocks?

```js
document.createTransition({
  updateDOM: () => new Promise(() => {}),
})
```

This might be done in code from LibraryA, but it means that all transitions are deadlocked from now on.

> 1. Execute the callbacks concurrently and while all steps for the transition are being executed. Since these callbacks are updating the DOM we could end up snapshotting a random frame and/or have the callbacks update the DOM midway through a transition. That seems pretty bad.

Here's where my head's at:

**Our API _enhances_ DOM updates with transitions.** Transitions aren't an essential part of the process, but the DOM update is. This is the thinking that led us to "the callback should always run, even if the transition cannot".

**Most DOM updates are synchronous, or queued behind a microtask.** In that case, we're the ones that make it async, due to the timing of the `updateDOM` callback. But even then, in these cases the DOM updates will complete in sequence.

**If the DOM updates are async regardless of transitions, the concurrency problem needs to be solved regardless of transitions.** If your async DOM framework breaks due to its own asyncness, then it isn't our responsibility to fix that. In fact, if we try and fix it, we're probably just getting in the way of the DOM frameworks that are correctly managing their own asyncness.

The Navigation API for instance doesn't provide a scheduler. It will allow many [intercept handlers](https://wicg.github.io/navigation-api/#ref-for-dom-navigateevent-intercept%E2%91%A0%E2%91%A3) to run concurrently. It provides an [abort signal](https://wicg.github.io/navigation-api/#ref-for-dom-navigateevent-signal%E2%91%A0), and trusts that the developer does the right thing to avoid overlapping actions.

I don't think we should provide a similar system, because skipping the transition doesn't always mean skipping the DOM change. It does in some cases, but that's judged by some other system (like the Navigation API).

So, if we call a second callback before the promise from the first has settled, it's only a problem if it would also be a problem without our API. So, it's not our problem to solve.

I might be missing part of the picture though!

Reply:
> This might be done in code from LibraryA, but it means that all transitions are deadlocked from now on.

Ah, so this would be the case of a developer error. We should limit the wait for the promise on a timeout, similar to what we do when dispatching this callback for the transition.

> If the DOM updates are async regardless of transitions, the concurrency problem needs to be solved regardless of transitions.

Let's say the developer is using the same callback outside of transitions. An event occurs which initiates the transition to state A and developer code issues the callback. This callback yields because of an async event (like disk fetch) and then another event happens which wants to transition to state B. The developer will already need to build a system to either queue the callback or handle the affect that they can be concurrent. That part SGTM.

But let's assume that the concurrency is handled by the developer code so its okay to run them in parallel. I think the developer would still want the DOM to reach a stable state, in this case to state A, before the browser snapshots the before state. If we go with the behaviour in 1) then they must build their own system to ensure all pending callbacks have finished executing before calling createTransition(...) again. With 2) as the default behaviour, if the developer doesn't want the pending callbacks to block then they can track the pending transition and use the skipTransition() API.

Reply:
> We should limit the wait for the promise on a timeout, similar to what we do when dispatching this callback for the transition.

But we can't be certain that it's 'stuck', it might just be slow. So, if we timeout, we could potentially be allowing two callbacks to run concurrently. And, if we're fine with that, why not just _always_ run them concurrently. That's more reliable for developers.

This is the case that made me change my mind about queuing.

> But let's assume that the concurrency is handled by the developer code so its okay to run them in parallel. I think the developer would still want the DOM to reach a stable state, in this case to state A, before the browser snapshots the before state. If we go with the behaviour in 1) then they must build their own system to ensure all pending callbacks have finished executing before calling createTransition(...) again. With 2) as the default behaviour, if the developer doesn't want the pending callbacks to block then they can track the pending transition and use the skipTransition() API.

So the code would be something like this:

```js
async function runUpdate() {
  const data = await fetchData();

  const transition = document.createTransition({
    async updateDOM() {
      await updateDOMSomehow(data);
      await loadFontsWithAgressiveTimeout();
    },
  });

  await transition.domUpdated;
}
```

Let's assume that the underlying framework can handle two concurrent DOM updates via `updateDOMSomehow`. I think it's fair to say it would only allow these to run concurrently if they're completely separate parts of the page. Otherwise, I'd expect the framework to either queue the `updateDOMSomehow` actions, or reject the first in favour of the last. So let's say the transitions are happening in ComponentA and ComponentB respectively.

The ideal here is probably two non-root transitions that can also run at the same time. But ignoring that…

The developer intends to perform two transitions. Snapping visual state and then transitioning is kinda ugly, so we're already in non-ideal territory. Given the lack of non-root transitions, the smoothest experience is probably to queue the transitions so each can run, but that isn't what the developer is doing in the code above (but our API does make it possible).

So any of the following could happen:

- Switch to ComponentA new state, transition to ComponentB state.
- Skip ComponentA new state transition mid-way, transition to ComponentB state.
- Transition for ComponentA is skipped, but it hasn't finished changing the DOM yet. Transition to ComponentB state. Mid-way through, ComponentA switches to new state.

Queuing the `updateDOM` callbacks prevents that third possibility. But, are we really preventing something undesirable?

Reply:
TODO:

- Add a note that changes are expected to be syncish, or the framework deals with it
- Don't need to get all transitions, because there'll be 0 or 1. Just store that on the doc.